# Análise Combinatória

## Importando biblioteca math

A biblioteca *math* só passou a possuir o método *comb* a partir do **Python 3.8**. Portanto, para funcionar, é preciso que o código esteja sendo executado nessa versão. Do contrário, irá ocorrer o erro **AttributeError: 'module' object has no attribute 'comb'**.

In [1]:
import math as mt

## Exemplo do boliche

Imagine que em um campo de boliche contém 5 bolas pretas e 6 bolas brancas, e a maquina de retorno de bolas apresenta aleatoriamente 2 bolas por vez, qual é a probabilidade de que uma bola sai branca e a outra preta?

In [2]:
n = 11  #número total de bolas
r = 2   #número selecionado de bolas

EA = mt.comb(n, r)
print("Espaço Amostral =", EA)

Espaço Amostral = 55


No caso acima, Espaço Amostral é a combinação de 2 em 2 dos 11 elementos que compõem o universo. O cálculo acima equivale a seguinte fórmula de combinação: **n!/(n-r)!*r!**

### Valores do epaço amostral

Para exemplificar as possíveis combinações, foi identificada cada bola pela numeração.

In [8]:
contador = 0;
strConcatenada = '| '
for x in range(1, 12): 
    for y in range(x, 12): 
        if (x != y) :
            contador = contador + 1;
            strConcatenada = strConcatenada + str(contador) + "(" + str(x) + "," + str(y) + ") | "
print(concatenador)

| 1(1,2) | 2(1,3) | 3(1,4) | 4(1,5) | 5(1,6) | 6(1,7) | 7(1,8) | 8(1,9) | 9(1,10) | 10(1,11) | 11(2,3) | 12(2,4) | 13(2,5) | 14(2,6) | 15(2,7) | 16(2,8) | 17(2,9) | 18(2,10) | 19(2,11) | 20(3,4) | 21(3,5) | 22(3,6) | 23(3,7) | 24(3,8) | 25(3,9) | 26(3,10) | 27(3,11) | 28(4,5) | 29(4,6) | 30(4,7) | 31(4,8) | 32(4,9) | 33(4,10) | 34(4,11) | 35(5,6) | 36(5,7) | 37(5,8) | 38(5,9) | 39(5,10) | 40(5,11) | 41(6,7) | 42(6,8) | 43(6,9) | 44(6,10) | 45(6,11) | 46(7,8) | 47(7,9) | 48(7,10) | 49(7,11) | 50(8,9) | 51(8,10) | 52(8,11) | 53(9,10) | 54(9,11) | 55(10,11) | 


### Selecionar as bolas

In [9]:
brancas = mt.comb(6,1)
pretas = mt.comb(5,1)

print("Chance de selcionar bolas brancas =", brancas)
print("Chance de selcionar bolas pretas =", pretas)

Chance de selcionar bolas brancas = 6
Chance de selcionar bolas pretas = 5


### Probabilidade de selecionar uma bola de cada cor

In [13]:
prob = (mt.comb(6,1) * mt.comb(5,1)) / mt.comb(11, 2)
prob = prob * 100
prob = round(prob, 2)

print("Probabilidade = ", prob, "%")

Probabilidade =  54.55 %


A fórmula de probabilidade é a seguinte: **(*'combinação de bolas brancas'* * *'combinação de bolas brancas'*) / *'combinação das 11 bolas 2 a 2'***. Chegando ao resultado de **54.55%** de chance de acontecer de sair uma bola de cada cor ao se retirar duas bolas das 11.